### PR TITLE
Add system spec for sync provider from find

### PR DESCRIPTION
**Pull request by davidgisbey**
## Context

The sync has some complex behaviour that's worth surfacing and making more obvious to any developers who have to work on it. One way of doing that is with a system spec.

Write a spec that covers how the sync handles:

- A site no longer being available for a given course between syncs.
- A course study_mode changing between syncs, and how this affects course options created in the first sync.

## Changes proposed in this pull request

- Add tests demonstrating the above behviour

## Guidance to review

Is this sitting in the right place? I'm not 100% sure where it should go?

## Link to Trello card

https://trello.com/c/adHlQT3h/1249-dev-add-system-spec-for-find-sync

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: spec/system/sync_provider_from_find/site_is_deleted_spec.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    given_a_site_is_deleted_which_is_not_associated_with_an_application
+
+    when_sync_provider_from_find_is_called
+    then_the_course_option_for_that_site_is_deleted
+
+    given_there_is_a_course_on_find_with_multiple_sites
+    and_sync_provider_from_find_has_been_called
+    and_raven_can_capture_exceptions
+
+    when_a_site_is_deleted_which_is_associted_to_an_application
```
Comment (tvararu) [inline spec/system/sync_provider_from_find/site_is_deleted_spec.rb:21]:
```suggestion
    when_a_site_is_deleted_which_is_associated_to_an_application
```

**Comment by duncanjbrown** [comment]
Love this. Should we document what happens when a part-time course option which has application choices associated with it sees its parent course go to full-time only? Does that option become "invalidated by Find" too?

File: spec/system/sync_provider_from_find/site_is_deleted_spec.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    SyncProviderFromFind.call(provider_name: 'ABC College', provider_code: 'ABC', sync_courses: true)
+  end
+
+  def then_the_correct_course_options_are_created_on_apply
+    @provider = Provider.find_by!(code: 'ABC')
+
+    expect(@provider.courses.first.course_options.count).to eq 2
+  end
+
+  def given_a_site_is_deleted_which_is_not_associated_with_an_application
```
Comment (malcolmbaig) [inline spec/system/sync_provider_from_find/site_is_deleted_spec.rb:41]: I think we should describe this slightly differently - current wording is a little ambiguous about whether we're referring to an apply Site record or a site returned in the Find API call.

Maybe something like `when_find_says_that_a_site_is_no_longer_listed_for_that_course` ?

File: spec/system/sync_provider_from_find/site_is_deleted_spec.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    when_sync_provider_from_find_is_called
+    then_the_course_option_for_that_site_is_deleted
+
+    given_there_is_a_course_on_find_with_multiple_sites
+    and_sync_provider_from_find_has_been_called
+    and_raven_can_capture_exceptions
+
+    when_a_site_is_deleted_which_is_associated_to_an_application
+    and_sync_provider_from_find_is_called
+    then_the_affected_course_options_invalidated_by_find_attriibute_is_set_to_true
```
Comment (malcolmbaig) [inline spec/system/sync_provider_from_find/site_is_deleted_spec.rb:23]: typo on attribute

**Comment by davidgisbey** [comment]
Currently, a course option's `invalidated_by_find` is not set to true if the parent course's study mode changes. The `invalidated_by_find` attribute is only used when a site has been deleted. When a study mode changes, invalidating a  course the vacancy status is set to 'no vacancies'.

File: spec/system/sync_provider_from_find/site_is_deleted_spec.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+  def then_the_course_option_for_that_site_is_deleted
+    expect(@provider.courses.first.course_options.count).to eq 1
+  end
+
+  def and_sync_provider_from_find_has_been_called
+    when_sync_provider_from_find_is_called
+  end
+
+  def when_a_site_is_deleted_which_is_associated_to_an_application
```
Comment (malcolmbaig) [inline spec/system/sync_provider_from_find/site_is_deleted_spec.rb:53]: Same wording observation as above, and perhaps move the application choice into a separate step.

So maybe:
```
when_find_says_that_a_site_is_no_longer_listed_for_that_course
and_that_site_is_part_of_an_application
```

File: spec/system/sync_provider_from_find/study_mode_changes_spec.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
+require 'rails_helper'
+
+RSpec.describe 'Sync from find' do
+  include FindAPIHelper
+
+  scenario 'a courses study mode changes between syncs' do
+    given_there_is_a_full_time_course_on_find
+
+    when_sync_provider_from_find_is_called
+    then_the_correct_course_option_are_created_on_apply
```
Comment (malcolmbaig) [inline spec/system/sync_provider_from_find/study_mode_changes_spec.rb:10]: `_is_created`?

**Comment by malcolmbaig** [comment]
Nicely done 👌 . Some small observations but overall LGTM.

File: spec/system/sync_provider_from_find/site_is_deleted_spec.rb (hunk 4 above)
Comment (davidgisbey) [reply spec/system/sync_provider_from_find/site_is_deleted_spec.rb]: Thanks for all the feedback

**Pull request merged**